### PR TITLE
Add available macros for 'return error code' as an alterative to 'verifyorexit'

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
Existing coding pattern of 'exit:' goto label is convenient for cleanup code, however it encourages C-style coding which has drawbacks:
   - less use of RAII (typical C++ cleanup)
   - forces variable declarations to be at the top of the scope due to the goto statement, moving variables further away from their usage.

 #### Summary of Changes
Creates Verify/return macros that are equivalent to the 'exit' macros and converts some files with sample usage for these macros.

Generated code is generally the same size as the previous goto code (with some slight immaterial improvements) and source code is shorter.

### CODE REVIEW

**Review by jelderton — APPROVED** [review]

**Review by rwalker-apple — APPROVED** [review]
yay!  on our way to no labels?

**Comment by andy31415** [comment]
> yay! on our way to no labels?

I am not a fan of `goto`. I understand it is useful for embedded, but it should not be forced on me. I would like everyone to have a choice.

File: src/lib/support/ReturnMacros.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+///   ReturnErrorOnFailure(channel->SendMsg(mnsg));
+#define ReturnErrorOnFailure(expr)                                                                                                 \
+    {                                                                                                                              \
+        CHIP_ERROR __err = (expr);                                                                                                 \
+        if (__err != CHIP_NO_ERROR)                                                                                                \
+        {                                                                                                                          \
+            return __err;                                                                                                          \
+        }                                                                                                                          \
+    }                                                                                                                              \
+    (void) 0
```
Comment (bzbarsky-apple) [inline src/lib/support/ReturnMacros.h:35]: I would strongly prefer we use:
```
do {
  ...
} while (0)
```
so the whole thing is in fact the single statement it looks like it is.  That way if people use it in `if ... else` without braces (which ideally they wouldn't do, I know), it will work correctly instead of giving confusing compiler errors.

Similar for the other macros defined here.

File: src/lib/support/ReturnMacros.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+#pragma once
+
+#include <core/CHIPError.h>
+
+/// Returns the error code if the expression returns something different
+/// than CHIP_NO_ERROR
+///
+/// Use like:
+///   ReturnErrorOnFailure(channel->SendMsg(mnsg));
```
Comment (bzbarsky-apple) [inline src/lib/support/ReturnMacros.h:26]: s/mnsg/msg/, please

File: src/lib/support/ReturnMacros.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    (void) 0
+
+/// Returns a specified error code if expression evaluates to true
+///
+/// Use like:
+///   ReturnErrorCodeIf(state == kInitialized, CHIP_NO_ERROR);
+///   ReturnErrorCodeIf(state == kInitialized, CHIP_ERROR_INVALID_STATE);
+#define ReturnErrorCodeIf(expr, code)                                                                                              \
+    {                                                                                                                              \
+        if ((expr))                                                                                                                \
```
Comment (bzbarsky-apple) [inline src/lib/support/ReturnMacros.h:57]: Are the extra parens here actually needed?

File: src/transport/RendezvousSession.cpp [hunk 4]
```diff
@@ -40,32 +41,29 @@ namespace chip {
 
 CHIP_ERROR RendezvousSession::Init(const RendezvousParameters & params)
 {
-    CHIP_ERROR err = CHIP_NO_ERROR;
-
     mParams = params;
-    VerifyOrExit(mDelegate != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
-    VerifyOrExit(mParams.HasSetupPINCode(), err = CHIP_ERROR_INVALID_ARGUMENT);
+    ReturnErrorCodeIf(mDelegate == nullptr, CHIP_ERROR_INCORRECT_STATE);
+    ReturnErrorCodeIf(!mParams.HasSetupPINCode(), CHIP_ERROR_INVALID_ARGUMENT);
```
Comment (bzbarsky-apple) [inline src/transport/RendezvousSession.cpp:46]: Why that and not `VerifyOrReturnError` here?  It's not clear to me whether there's a principled way to decide between these two macros...

File: src/transport/raw/TCP.cpp [hunk 5]
```diff
@@ -162,75 +165,58 @@ Inet::TCPEndPoint * TCPBase::FindActiveConnection(const PeerAddress & address)
...
-
-    endPoint = FindActiveConnection(address);
+    Inet::TCPEndPoint * endPoint = FindActiveConnection(address);
 
     if (endPoint != nullptr)
     {
-        err = endPoint->Send(msgBuf);
+        return endPoint->Send(autofree.Release());
     }
     else
```
Comment (bzbarsky-apple) [inline src/transport/raw/TCP.cpp:216]: What's our general position on else after return?

Comment (gerickson) [reply src/transport/raw/TCP.cpp:216]: I generally prefer single point of return or exit / error handling (RAII or not) which moots this particular facet of coding style discussion.

Comment (andy31415) [reply src/transport/raw/TCP.cpp:216]: I have created this PR after seeing that mandating goto with single point of exit can result in significantly hard to follow code. 

The premise is that we will have the option to use return if the code is cleaner as a result. Mandating exit label always seems too strong of a requirement.

File: src/transport/RendezvousSession.cpp (hunk 4 above)
Comment (andy31415) [reply src/transport/RendezvousSession.cpp:46]: I have no clear way to differentiate between the two, I changed this as an example.

I believe both are readable. My best guess of choosing would be 'negations are hard to follow' so the correct way here seems indeed VerifyOrReturnError because HasSetupPinCode is otherwise negated. I will update the code.

File: src/transport/raw/TCP.cpp (hunk 5 above)
Comment (andy31415) [reply src/transport/raw/TCP.cpp:216]: Regarding the else after return: I think generally I would say no else is needed after return. In this particular case however, this is a final statement that reads like:

```cpp
return (endpoint != nullptr)
    ? endPoint->Send(autofree.Release())
    : SendAfterConnect(address, autofree.Release());
```

so I thought else helps clarity. I would drop the else if the readability reviews would ask for it.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/transport/raw/TCP.cpp (hunk 5 above)
Comment (andy31415) [reply src/transport/raw/TCP.cpp:216]: +1 that mixing in a single method would be absolutely bad - you get the benefit of neither. I am not arguing for that. I am arguing that I should be allowed to use return if it makes the code easier/clearer to read.

Comment (andy31415) [reply src/transport/raw/TCP.cpp:216]: I could not find a good way to count, but `find . -type f | egrep -v '/third_party/' | xargs egrep -zc 'exit:\s*return' | grep -v ':0' | wc -l` seems to indicate that there are 149 instances today where 'exit: return...' is used. I believe those instances would benefit from just using return in the first place.

Comment (mspang) [reply src/transport/raw/TCP.cpp:216]: C forces you into the goto exit pattern, but C++ doesn't, and the goto style with declarations front loaded seems unusual for C++. Google's style guide doesn't even bother to recommend against using goto because it is so uncommon in the first place [1].

[1] https://google.github.io/styleguide/cppguide.html#Goals

**Review by mspang — APPROVED** [review]
